Override memory limits

Jw0x47 opened the pull request:
Cc: @gchomatas

Commit:
Allow resource overrides in build configs.

Commit:
tests

Commit:
move this back

Commit:
change mysql query to save config during enqueue

Commit:
provide implementations of equals and hashcode for object that goes into a set

gchomatas commented on BlazarBase/src/main/java/com/hubspot/blazar/base/BuildConfig.java:
```diff
@@ -103,8 +112,211 @@ public static BuildConfig makeDefaultBuildConfig(){
+      return new BuildConfig(steps, before, after, env, buildDeps, webhooks, cache, buildpack, user, stepActivation, buildResources, depends, provides);
+    }
+
+    public List<BuildStep> getSteps() {
```
It is not a good practice to have getters in the builder. This will allow you to pass around the builder instead of the real object which essentially means that you created a fully mutable object disguised as a builder.

gchomatas commented on BlazarBase/src/main/java/com/hubspot/blazar/base/StepActivationCriteria.java:
```diff
@@ -1,11 +1,12 @@
 
-import java.util.Set;
-
 public class StepActivationCriteria implements Predicate<GitInfo> {
```
This is quite a misleading name. What about ContainsBranchCriterion?

gchomatas commented on BlazarBase/src/main/java/com/hubspot/blazar/external/models/singularity/Resources.java:
```diff
@@ -0,0 +1,42 @@
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class Resources {
```
DockerContainerResources or if you want to be more generic BuildProcessContainerResources

gchomatas commented on BlazarBase/src/main/java/com/hubspot/blazar/external/models/singularity/Resources.java:
```diff
@@ -0,0 +1,42 @@
+  private final double cpus;
+  private final double memoryMb;
+
+  public static com.hubspot.mesos.Resources toMesosResources(Resources resources) {
```
Why is this a static methods? If you already have a BuildProcessContainerResources instance you can just do: buildProcessContainerResources.toMesosResoures.toMesosResources();

gchomatas commented on BlazarBase/src/main/java/com/hubspot/blazar/external/models/singularity/Resources.java:
```diff
@@ -0,0 +1,42 @@
+    return new Resources(resources.getCpus(), resources.getMemoryMb());
+  }
+
+  public static Resources add(Resources a, Resources b) {
```
Same here you don't need static: resource.add(anotherResource)
Also implement the add in the class instead of using the mesos class facility and converting back to your class. It is not a good practice to use the implementation of a class  you don't control and it is not meant to be a lib class with a well defined interface. The implementation can change at any time and your add method will change behavior without you having a way to know.

gchomatas commented on BlazarData/src/main/java/com/hubspot/blazar/data/dao/ModuleBuildDao.java:
```diff
@@ -69,10 +69,10 @@
 
   @GetGeneratedKeys
-  @SqlUpdate("INSERT INTO module_builds (repoBuildId, moduleId, buildNumber, state) VALUES (:repoBuildId, :moduleId, :buildNumber, :state)")
+  @SqlUpdate("INSERT INTO module_builds (repoBuildId, moduleId, buildNumber, state, buildConfig, resolvedConfig) VALUES (:repoBuildId, :moduleId, :buildNumber, :state, :buildConfig, :resolvedConfig)")
```
Better name them to buildConfig, resolvedBuildConfig

gchomatas commented on BlazarData/src/main/java/com/hubspot/blazar/data/service/ModuleBuildService.java:
```diff
@@ -104,10 +105,15 @@ public void skip(RepositoryBuild repositoryBuild, Module module) {
   }
 
-  public void enqueue(RepositoryBuild repositoryBuild, Module module) {
+  public void enqueue(RepositoryBuild repositoryBuild, Module module, BuildConfig baseConfig, BuildConfig resolvedConfig) {
```
enqueue(RepositoryBuild repositoryBuild, Module module, BuildConfig buildConfig, BuildConfig resolvedBuildConfig)

gchomatas commented on BlazarData/src/main/java/com/hubspot/blazar/data/service/ModuleBuildService.java:
```diff
@@ -104,10 +105,15 @@ public void skip(RepositoryBuild repositoryBuild, Module module) {
     int nextBuildNumber = repositoryBuild.getBuildNumber();
     LOG.info("Enqueuing build for module {} with build number {}", module.getId().get(), nextBuildNumber);
-    ModuleBuild build = ModuleBuild.queuedBuild(repositoryBuild, module, nextBuildNumber);
+    ModuleBuild build =
```
Either change the static method (`queuedBuild`) to use the new args or set all required args in a single builder, i.e. explicitly set the State.QUEUED in the builder.

gchomatas commented on BlazarService/src/main/java/com/hubspot/blazar/config/ExecutorConfiguration.java:
```diff
@@ -16,16 +18,22 @@
 
   public static ExecutorConfiguration defaultConfiguration() {
-    return new ExecutorConfiguration(null, null);
+    return new ExecutorConfiguration(null, null, null);
```
ExecutorConfiguration(Optional.absent(), Optional.absent(), Optional.absent());

gchomatas commented on BlazarService/src/main/java/com/hubspot/blazar/listener/LaunchingRepositoryBuildVisitor.java:
```diff
@@ -197,4 +210,24 @@ private boolean lastBuildSucceeded(Module module) {
+    String sha = branchBuild.getCommitInfo().get().getCurrent().getId();
+    String folder = module.getFolder();
+    String configPath = folder + (folder.isEmpty() ? "" : "/") + ".blazar.yaml";
+    BuildConfig baseConfig = buildConfigUtils.getConfigAtRefOrDefault(branch, configPath, sha);
```
I don't see any try/catch in the `visitLaunching` method. What will happen if this method throws and error?

Jw0x47 replied:
They will be retried by the stateful event bus so long as the exception is not a `NonRetryableBuildException`.
https://github.com/HubSpot/Blazar/blob/override-memory-limits/BlazarService/src/main/java/com/hubspot/blazar/listener/BuildEventDispatcher.java#L55

gchomatas commented on BlazarService/src/main/java/com/hubspot/blazar/util/BuildConfigUtils.java:
```diff
@@ -0,0 +1,117 @@
+    return getConfigAtRefOrDefault(gitInfo, BUILDPACK_FILE, gitInfo.getBranch());
+  }
+
+  public BuildConfig ensureDefaultConfigurationForBuild(BuildConfig buildConfig) {
```
I think `addMissingBuildConfigSettings` could be more easy to understand than `ensureDefaultConfigurationForBuild`

gchomatas commented on BlazarService/src/main/java/com/hubspot/blazar/util/BuildConfigUtils.java:
```diff
@@ -0,0 +1,117 @@
+    return buildConfigWithDefaults.build();
+  }
+
+  public BuildConfig mergeConfig(BuildConfig primary, BuildConfig secondary) {
```
Add some comments to explain how merge is implemented and why this unconventional mix of substitution of some settings and merge of some others has been chosen.

gchomatas commented on BlazarService/src/test/java/com/hubspot/blazar/util/BlazarConfigUtilsTest.java:
```diff
@@ -0,0 +1,138 @@
+  private static final GitInfo PRIMARY_BRANCH = GitInfo.fromString("git.example.com/TestOrg/repo.git#primary");
+  private static final GitInfo SECONDARY_BRANCH = GitInfo.fromString("git.example.com/TestOrg/repo.git#secondary");
+
+  private final GitHubHelper gitHubHelper = mock(GitHubHelper.class);
```
All these mocked instances could be static

gchomatas commented on BlazarService/src/test/java/com/hubspot/blazar/util/BlazarConfigUtilsTest.java:
```diff
@@ -0,0 +1,138 @@
+
+  @Test
+  public void itPrefersPrimaryConfigForNonMergableFieldsWhenMergingConfigs() {
+    BuildConfigUtils configUtils = new BuildConfigUtils(blazarConfiguration, gitHubHelper);
```
Refactor it to a static and reuse it in all tests

gchomatas commented on BlazarService/src/test/java/com/hubspot/blazar/util/BlazarConfigUtilsTest.java:
```diff
@@ -0,0 +1,138 @@
+import com.hubspot.blazar.exception.NonRetryableBuildException;
+import com.hubspot.blazar.external.models.singularity.Resources;
+
+@RunWith(JukitoRunner.class)
```
You manually mock and you don't inject anything so you don't need JukitoRunner

Commit:
Changes per comments on PR

- Remove getters from builders
- Make mocks in new test static, remove unused runner
- Document merge of build configs
- Rename some methods / vars for clarity

Commit:
fix caps

Jw0x47 committed:
Delete this conflicting file.

Case sensitive file systems are cool.

Commit:
Merge branch 'hs-master' into override-memory-limits

gchomatas commented:
🚢
